sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -68,7 +68,13 @@ case class EnsureRequirements(
       case (child, distribution) =>
         val numPartitions = distribution.requiredNumPartitions
           .getOrElse(conf.numShufflePartitions)
-        ShuffleExchangeExec(distribution.createPartitioning(numPartitions), child, shuffleOrigin)
+        // Avoid generating redundant ShuffleExchangeExec node
+        val shuffleChild = child match {
+          case s: ShuffleExchangeExec => s.child
+          case other => other
+        }
+        ShuffleExchangeExec(
+          distribution.createPartitioning(numPartitions), shuffleChild, shuffleOrigin)
```

Comment:
@ulysses-you It only takes effect if there are no operators between two shuffles.

Reply:
My point is that, it's a user-specified shuffle, we can not just remove it. Does it affect shuffle reuse ? e.g.,

```
with t (select /*+ repartition */ * from t)
select * from t join t2
union all
select * from t
```

Reply:
Skip reuse shuffle case.